File: src/com/suushiemaniac/cubing/alglib/lang/SquareOneAlgorithmReader.java
package com.suushiemaniac.cubing.alglib.lang;

import com.suushiemaniac.cubing.alglib.alg.Algorithm;
import com.suushiemaniac.cubing.alglib.alg.SimpleAlg;
import com.suushiemaniac.cubing.alglib.alg.commutator.Commutator;
import com.suushiemaniac.cubing.alglib.alg.commutator.PureComm;
import com.suushiemaniac.cubing.alglib.alg.commutator.SetupComm;
import com.suushiemaniac.cubing.alglib.exception.InvalidNotationException;
import com.suushiemaniac.cubing.alglib.lang.antlr.squareone.SquareOneBaseVisitor;
import com.suushiemaniac.cubing.alglib.lang.antlr.squareone.SquareOneLexer;
import com.suushiemaniac.cubing.alglib.lang.antlr.squareone.SquareOneParser;
import com.suushiemaniac.cubing.alglib.move.SquareOneMove;
import com.suushiemaniac.cubing.alglib.move.modifier.SquareOneDirectionModifier;
import com.suushiemaniac.cubing.alglib.move.modifier.SquareOneHalfModifier;
import com.suushiemaniac.cubing.alglib.move.modifier.SquareOneModifier;
import com.suushiemaniac.cubing.alglib.move.modifier.SquareOneNumModifier;
import org.antlr.v4.runtime.*;
import org.antlr.v4.runtime.tree.ParseTree;

public class SquareOneAlgorithmReader extends SquareOneBaseVisitor<Algorithm> implements NotationReader {
    @Override
    public Algorithm parse(String input) {
        SquareOneLexer lexer = new SquareOneLexer(new ANTLRInputStream(input));
        CommonTokenStream tokens = new CommonTokenStream(lexer);
        SquareOneParser parser = new SquareOneParser(tokens);
        parser.removeErrorListeners();
        parser.addErrorListener(new BaseErrorListener() {
            public void syntaxError(Recognizer<?, ?> recognizer, Object offendingSymbol, int line, int charPositionInLine, String msg, RecognitionException e) {
                throw new InvalidNotationException(input, msg);
            }
        });
        ParseTree tree = parser.squareOne();
        return this.visit(tree);
    }

    private class SquareOneMoveReader extends SquareOneBaseVisitor<SquareOneMove> {
        private SquareOneModifierReader modifierReader;

        public SquareOneMoveReader(SquareOneModifierReader modifierReader) {
            this.modifierReader = modifierReader;
        }

        @Override
        public SquareOneMove visitSquareOneMoveSlash(SquareOneParser.SquareOneMoveSlashContext ctx) {
            SquareOneModifier modifier = this.modifierReader.visit(ctx.squareOneModifier());
            boolean endSlash = ctx.SQUARE_ONE_SLASH() != null;
            return new SquareOneMove(modifier, false, endSlash);
        }
    }

    private class SquareOneModifierReader extends SquareOneBaseVisitor<SquareOneModifier> {
        private SquareOneHalfModifierReader halfModifierReader;

        public SquareOneModifierReader() {
            this.halfModifierReader = new SquareOneHalfModifierReader();
        }

        @Override
        public SquareOneModifier visitSquareOneModifier(SquareOneParser.SquareOneModifierContext ctx) {
            SquareOneHalfModifier halfUp = this.halfModifierReader.visit(ctx.squareOneFaceTurn(0));
            SquareOneHalfModifier halfDown = this.halfModifierReader.visit(ctx.squareOneFaceTurn(1));
            return new SquareOneModifier(halfUp, halfDown);
        }
    }

    private class SquareOneHalfModifierReader extends SquareOneBaseVisitor<SquareOneHalfModifier> {
        @Override
        public SquareOneHalfModifier visitSquareOneFaceTurn(SquareOneParser.SquareOneFaceTurnContext ctx) {
            SquareOneDirectionModifier directionModifier = SquareOneDirectionModifier.fromNotation(ctx.SQUARE_ONE_PLANE_ZERO() == null ? (ctx.SQUARE_ONE_DIRECTION_MODIFIER() == null ? "" : ctx.SQUARE_ONE_DIRECTION_MODIFIER().getText()) : "");
            SquareOneNumModifier halfModifier = SquareOneNumModifier.fromNotation(ctx.SQUARE_ONE_PLANE_ZERO() == null ? ctx.SQUARE_ONE_PLANE_POS().getText() : "0");
            return new SquareOneHalfModifier(directionModifier, halfModifier);
        }
    }

    private class SquareOneCommReader extends SquareOneBaseVisitor<Commutator> {
        private SquareOneAlgorithmReader algorithmReader;

        public SquareOneCommReader(SquareOneAlgorithmReader algorithmReader) {
            this.algorithmReader = algorithmReader;
        }

        @Override
        public Commutator visitSquareOnePureComm(SquareOneParser.SquareOnePureCommContext ctx) {
            Algorithm partA = this.algorithmReader.visit(ctx.squareOne(0));
            Algorithm partB = this.algorithmReader.visit(ctx.squareOne(1));
            return new PureComm(partA, partB);
        }

        @Override
        public Commutator visitSquareOneSetupComm(SquareOneParser.SquareOneSetupCommContext ctx) {
            Algorithm partA = this.algorithmReader.visit(ctx.squareOne(0));
            Algorithm partB = this.algorithmReader.visit(ctx.squareOne(1));
            return new SetupComm(partA, partB);
        }
    }

    private SquareOneModifierReader modifierReader;
    private SquareOneMoveReader moveReader;
    private SquareOneCommReader commReader;

    public SquareOneAlgorithmReader() {
        this.modifierReader = new SquareOneModifierReader();
        this.moveReader = new SquareOneMoveReader(this.modifierReader);
        this.commReader = new SquareOneCommReader(this);
    }

    @Override
    public SimpleAlg visitSquareOneAlg(SquareOneParser.SquareOneAlgContext ctx) {
        SquareOneMove[] moves = new SquareOneMove[ctx.squareOneMoveSlash().size() + 1];
        SquareOneModifier firstModifier = this.modifierReader.visit(ctx.squareOneMoveSlash().size() > 0 ? ctx.squareOneMoveSlash(0).squareOneModifier() : ctx.squareOneModifier());
        boolean firstBeginSlash = ctx.squareOneBeginSlash() != null;
        boolean firstEndSlash = (ctx.squareOneMoveSlash().size() > 0 ? ctx.squareOneMoveSlash(0).SQUARE_ONE_SLASH() : ctx.SQUARE_ONE_SLASH()) != null;
        SquareOneMove firstMove = new SquareOneMove(firstModifier, firstBeginSlash, firstEndSlash);
        moves[0] = firstMove;
        if (ctx.squareOneMoveSlash().size() < 1) return new SimpleAlg(moves);
        if (ctx.squareOneMoveSlash().size() > 1)
            for (int i = 1; i < ctx.squareOneMoveSlash().size(); i++) moves[i] = this.moveReader.visit(ctx.squareOneMoveSlash(i));
        SquareOneModifier lastModifier = this.modifierReader.visit(ctx.squareOneModifier());
        boolean lastEndSlash = ctx.SQUARE_ONE_SLASH() != null;
        moves[moves.length - 1] = new SquareOneMove(lastModifier, false, lastEndSlash);
        return new SimpleAlg(moves);
    }

    @Override
    public Commutator visitSquareOneComm(SquareOneParser.SquareOneCommContext ctx) {
        return this.commReader.visit(ctx);
    }
}

File: src/main/kotlin/com/suushiemaniac/cubing/alglib/lang/SquareOneAlgorithmReader.kt
package com.suushiemaniac.cubing.alglib.lang

import com.suushiemaniac.cubing.alglib.alg.Algorithm
import com.suushiemaniac.cubing.alglib.alg.SimpleAlg
import com.suushiemaniac.cubing.alglib.alg.commutator.Commutator
import com.suushiemaniac.cubing.alglib.alg.commutator.PureComm
import com.suushiemaniac.cubing.alglib.alg.commutator.SetupComm
import com.suushiemaniac.cubing.alglib.antlr.SquareOneBaseVisitor
import com.suushiemaniac.cubing.alglib.antlr.SquareOneLexer
import com.suushiemaniac.cubing.alglib.antlr.SquareOneParser
import com.suushiemaniac.cubing.alglib.move.Move
import com.suushiemaniac.cubing.alglib.move.SquareOneMove
import com.suushiemaniac.cubing.alglib.move.modifier.SquareOneDirectionModifier
import com.suushiemaniac.cubing.alglib.move.modifier.SquareOneHalfModifier
import com.suushiemaniac.cubing.alglib.move.modifier.SquareOneModifier
import com.suushiemaniac.cubing.alglib.move.modifier.SquareOneNumModifier
import com.suushiemaniac.cubing.alglib.util.ParseUtils.fromNotation

class SquareOneAlgorithmReader : SquareOneBaseVisitor<Algorithm>(), NotationReader {
    private val modifierReader = SquareOneModifierReader()

    private val moveReader = SquareOneMoveReader(this.modifierReader)
    private val commReader = SquareOneCommReader(this)

    override fun parse(input: String): Algorithm {
        return NotationReader.parseString(input, ::SquareOneLexer, ::SquareOneParser, SquareOneParser::squareOne, this::visit)
    }

    private inner class SquareOneMoveReader(private val modifierReader: SquareOneModifierReader) : SquareOneBaseVisitor<SquareOneMove>() {
        override fun visitSquareOneModifierSlash(ctx: SquareOneParser.SquareOneModifierSlashContext): SquareOneMove {
            val modifier = this.modifierReader.visit(ctx.squareOneModifier())

            return SquareOneMove(modifier, isBeginSlash = false, isEndSlash = true)
        }
    }

    private inner class SquareOneModifierReader : SquareOneBaseVisitor<SquareOneModifier>() {
        private val halfModifierReader = SquareOneHalfModifierReader()

        override fun visitSquareOneModifier(ctx: SquareOneParser.SquareOneModifierContext): SquareOneModifier {
            val halfUp = this.halfModifierReader.visit(ctx.squareOneFaceTurn(0))
            val halfDown = this.halfModifierReader.visit(ctx.squareOneFaceTurn(1))

            return SquareOneModifier(halfUp, halfDown)
        }
    }

    private inner class SquareOneHalfModifierReader : SquareOneBaseVisitor<SquareOneHalfModifier>() {
        override fun visitSquareOneFaceTurn(ctx: SquareOneParser.SquareOneFaceTurnContext): SquareOneHalfModifier {
            val directNotation = if (ctx.SQUARE_ONE_PLANE_ZERO() == null) if (ctx.SQUARE_ONE_DIRECTION_MODIFIER() == null) "" else ctx.SQUARE_ONE_DIRECTION_MODIFIER().text else ""
            val directionModifier = SquareOneDirectionModifier.values().fromNotation(directNotation)

            val halfNotation = if (ctx.SQUARE_ONE_PLANE_ZERO() == null) ctx.SQUARE_ONE_PLANE_POS().text else "0"
            val halfModifier = SquareOneNumModifier.values().fromNotation(halfNotation)

            return SquareOneHalfModifier(directionModifier, halfModifier)
        }
    }

    private inner class SquareOneCommReader(private val algorithmReader: SquareOneAlgorithmReader) : SquareOneBaseVisitor<Commutator>() {
        override fun visitSquareOnePureComm(ctx: SquareOneParser.SquareOnePureCommContext): Commutator {
            val partA = this.algorithmReader.visit(ctx.squareOneAlg(0))
            val partB = this.algorithmReader.visit(ctx.squareOneAlg(1))

            return PureComm(partA, partB)
        }

        override fun visitSquareOneSetupComm(ctx: SquareOneParser.SquareOneSetupCommContext): Commutator {
            val partA = this.algorithmReader.visit(ctx.squareOneAlg(0))
            val partB = this.algorithmReader.visit(ctx.squareOneAlg(1))

            return SetupComm(partA, partB)
        }
    }

    override fun visitSquareOne(ctx: SquareOneParser.SquareOneContext): Algorithm {
        return if (ctx.squareOneAlg() != null) this.visit(ctx.squareOneAlg()) else SimpleAlg()
    }

    override fun visitSquareOneSimple(ctx: SquareOneParser.SquareOneSimpleContext): SimpleAlg {
        val seqMoves = ctx.squareOneModifierSlash().map(this.moveReader::visit) +
                SquareOneMove(this.modifierReader.visit(ctx.squareOneModifier()), false, ctx.squareOneEndSlash() != null)

        val algMoves = listOf(seqMoves.first().copy(isBeginSlash = ctx.squareOneBeginSlash() != null)) +
                seqMoves.drop(1)

        return SimpleAlg(algMoves)
    }

    override fun visitSquareOneComm(ctx: SquareOneParser.SquareOneCommContext): Commutator {
        return this.commReader.visit(ctx)
    }
}